### PR TITLE
Optional prepareObject argument for makeHardener

### PR DESCRIPTION
**Pull request by michaelfig**
This optional argument allows makeHardener to be adapted to more subtle shallow object freezing rules, without compromising its simplicity or needing to change Object.freeze itself.

I will write tests when I have something working.

### CODE REVIEW

**Review by erights — COMMENTED** [review]
Hi Michael, I don't see anything wrong[1] with this PR, but I also don't see the motivation. Do you have a use case in mind?

[1] Via the EcmaScript object invariants, `Object.freeze` enforces some stability properties. The order in which `harden` freezes and traverses leverages these stability properties to ensure that the traversed own properties as seen during harden cannot change after harden. This enables us to state simply what guarantees a successful harden provides.

This is not an argument against this PR. But it does make the semantics of `makeHardener` harder to explain. So if there is no problem that this PR would address, I'm skeptical of it.

File: src/index.js [hunk 1]
```diff
@@ -60,6 +54,14 @@ function makeHardener(initialFringe, freeze) {
     }
 
     function freezeAndTraverse(obj) {
+      // Naively prepare the object for freezing.
```
Comment (erights) [inline src/index.js:57]: Now that I know the use case you have in mind, let's discuss `immunize` further before revisiting this PR. I remain skeptical that `immunize` is a good idea. (Great name btw ;).) Example:

```js
function foo() {}
const dag = { foo, foo };
// dag.foo === dag.foo

immunize(dag);
// dag.foo !== dag.foo
```

Even if you complexify `immunize` enough to handle such cases, I remain nervous about trying to do surgery to the object graph in place, especially transitively.

Comment (michaelfig) [reply src/index.js:57]: Let's continue this in [the Jessie issue](https://github.com/Agoric/Jessie/issues/27).

**Comment by michaelfig** [comment]
Hi @erights, I'm working on modifying https://github.com/Agoric/SES to allow specifying custom whitelists (i.e. use the Jessie whitelist), and this feature is the only thing standing in my way of being able to test Jessie code with an `immunize()` endowment, as it needs to use the same fringe as the original harden or everything breaks.

I could use my repo's fork, but I would greatly appreciate having this upstream so my sources don't drift from the mainline.

Thanks,
Michael.

**Comment by michaelfig** [comment]
I'm closing this for now... I need to rethink the interface I need, and will open a new pr when I have something.

**Comment by michaelfig** [comment]
I've added an optional `opts` argument, and unit tests for specifying a `fringeSet` (so that multiple hardeners can share the same set), and also a `naivePrepareObject` (for just this single hardener), as the old PR had.

**Comment by michaelfig** [comment]
Continuing in #35.

**Pull request closed without merging**